pymodbus/client/base.py

```diff
@@ -4,9 +4,9 @@
 import asyncio
 import socket
 from abc import abstractmethod
-from collections.abc import Awaitable, Callable
+from collections.abc import Callable, Coroutine
 from dataclasses import dataclass
-from typing import Any, cast
+from typing import Any, Union, cast
```

Comment:
Use `from __future__ import annotations` to match the rest of the codebase.

This allows using `X | Y` for instead of `Union[X | Y]` [Union types ](https://docs.python.org/3/library/stdtypes.html#union-type)

Reply:
> Use `from __future__ import annotations` to match the rest of the codebase.

https://github.com/pymodbus-dev/pymodbus/blob/72e3399d41c20719ac3e63412fd1be2b8cca3840/pymodbus/client/base.py#L2

That import already exists.

> This allows using `X | Y` for instead of `Union[X | Y]` [Union types ](https://docs.python.org/3/library/stdtypes.html#union-type)

Not sure why exactly, but the normal style doesn't seem to work here.

Reply:
??

Works for me with both `pyright` and `mypy`.  Did you mismatch a bracket, perhaps?

![image](https://github.com/user-attachments/assets/b19d5e92-e790-426d-a54c-c219c5412eb5)

Reply:
> Works for me with both `pyright` and `mypy`. Did you mismatch a bracket, perhaps?

Oh, the error is coming from `pylint`, not `pyright` or `mypy`:
```python
$ pylint --recursive=y examples pymodbus test
************* Module pymodbus.client.base
pymodbus/client/base.py:23:61: E1131: unsupported operand type(s) for | (unsupported-binary-operation)
pymodbus/client/base.py:228:45: E1131: unsupported operand type(s) for | (unsupported-binary-operation)
```

Reply:
Bah, appears to be a `pylint` bug.  https://github.com/pylint-dev/pylint/issues/7381.

Use `# pylint: disable=unsupported-binary-operation`, I guess.

Reply:
Hmm, the normal `|` style seems to be entirely broken on python 3.9:
```python
ERROR test/framers/test_old_framers.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/framers/test_old_framers.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/framers/test_old_framers.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_client/test_client.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_client/test_client.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_client/test_client.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_client/test_client_sync.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_client/test_client_sync.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_client/test_client_sync.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_examples/test_client_server_async.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_examples/test_client_server_async.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_examples/test_client_server_async.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_examples/test_client_server_sync.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_examples/test_client_server_sync.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_examples/test_client_server_sync.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_examples/test_examples.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_examples/test_examples.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/sub_examples/test_examples.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/test_network.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/test_network.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
ERROR test/test_network.py - TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
```

Reply:
To date, the test have not had type hints (since it seemed like low value).  No problem with adding them, but in this case you will need to add `from __future__ import annotations` for the first time.

Reply:
> No problem with adding them, but in this case you will need to add `from __future__ import annotations` for the first time.

Tried this, doesn't seem to help, looks to just be a limitation of python 3.9.

Reply:
I agree with @alexrudd2 adding typing to tests, hav3 no real value and will most likely just complicate writing tests.

Please do not go down that route.

Reply:
> adding typing to tests, hav3 no real value and will most likely just complicate writing tests.

So I'm not really trying to add types to the tests so much as make it so that we don't get type errors when we typecheck the tests.

Reply:
> typecchecking tests is an unwanted feature.

I'm still confused about this, I mean if the tests have incorrect types then they are likely to cause issues.

> Even if we were to consider it, you should NOT mix production code changes with test changes.

Yeah, going to split things out more.